### PR TITLE
Resolve JS-6 Implement S5260 in js/ts

### PR DESCRIPTION
**Pull request by zglicz**
Reusing most of the logic from S5256 to compute the grid structure. Taking that logic to a helper file about tables.

### CODE REVIEW

File: packages/jsts/src/rules/helpers/table.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  const spanAttr = getProp(tree.openingElement.attributes, spanKey);
+  if (spanAttr) {
+    span = parseInt(String(getLiteralPropValue(spanAttr)));
+  }
+  return span;
+}
+
+function rowSpan(tree: TSESTree.JSXElement): number {
+  let value = computeSpan(tree, 'rowspan');
+  const MAX_ROW_SPAN = 65534;
```
Comment (vdiez) [inline packages/jsts/src/rules/helpers/table.ts:50]: no need to declare const on each call, you can move this to the top of the file.

File: packages/jsts/src/rules/helpers/table.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  if (value > MAX_ROW_SPAN) {
+    value = MAX_ROW_SPAN;
+  }
+  return value;
+}
+
+function colSpan(tree: TSESTree.JSXElement): number {
+  let value = computeSpan(tree, 'colspan');
+  const MAX_INVALID_COL_SPAN = 10000;
```
Comment (vdiez) [inline packages/jsts/src/rules/helpers/table.ts:59]: same

File: packages/jsts/src/rules/helpers/table.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+  tree.children.forEach(child => {
+    if (child.type === 'JSXElement') {
+      const childType = getElementType(context)(child.openingElement).toLowerCase();
+      if (childType === 'tr') {
+        rows.push(extractRow(child));
+      } else if (childType === 'table') {
+        // skip
+      } else {
+        const KNOWN_TABLE_STRUCTURE_ELEMENTS = ['thead', 'tbody', 'tfoot'];
```
Comment (vdiez) [inline packages/jsts/src/rules/helpers/table.ts:137]: same, move this declaration to top of the page

File: packages/jsts/src/rules/helpers/table.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        if (KNOWN_TABLE_STRUCTURE_ELEMENTS.includes(childType)) {
+          handleInternalStructure(child);
```
Comment (vdiez) [inline packages/jsts/src/rules/helpers/table.ts:139]: why not calling direclty extractRows? don't understand the need for `handleInternalStructure`

File: packages/jsts/src/rules/helpers/table.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+          unknownTableStructure = true;
+        }
+      }
+    } else if (
+      child.type === 'JSXExpressionContainer' &&
+      child.expression.type !== 'JSXEmptyExpression'
+    ) {
+      unknownTableStructure = true;
+    } else if (child.type === 'JSXFragment') {
+      handleInternalStructure(child);
```
Comment (vdiez) [inline packages/jsts/src/rules/helpers/table.ts:150]: same, why not calling directly extractRows?

**Review by vdiez — CHANGES_REQUESTED** [review]

File: packages/jsts/src/rules/helpers/table.ts (hunk 4 above)
Comment (zglicz) [reply packages/jsts/src/rules/helpers/table.ts:139]: as in to do what?
  rows.concat(extractRows) ?
We need to perform the check if its null to know if its an unknown structure. I do believe we need to run the logic from handleInternalStructure to correctly propagate the unknown structure or add the rows.

File: packages/jsts/src/rules/helpers/isHtmlElement.ts [hunk 6]
```diff
@@ -175,6 +175,6 @@ const HTML_TAG_NAMES = new Set([
 export function isHtmlElement(node: TSESTree.JSXElement) {
   return (
     node.openingElement.name.type === 'JSXIdentifier' &&
-    HTML_TAG_NAMES.has(node.openingElement.name.name.toLowerCase())
+    HTML_TAG_NAMES.has(node.openingElement.name.name)
```
Comment (vdiez) [inline packages/jsts/src/rules/helpers/isHtmlElement.ts:178]: why remove the lowercase comparison?

Comment (zglicz) [reply packages/jsts/src/rules/helpers/isHtmlElement.ts:178]: While checking on peach, I found this code snippet - https://peach.aws-prd.sonarsource.com/issues?projects=js%3Agutenberg&resolved=false&rules=javascript%3AS5256&open=AY70ITYshdJEwEPe_w_q

JSX is case sensitive and we can differentiate between, in this case, <Section> vs <section>

`<table
	className={ classes === '' ? undefined : classes }
	style={ { ...colorProps.style, ...borderProps.style } }
>
	<Section type="head" rows={ head } />
	<Section type="body" rows={ body } />
	<Section type="foot" rows={ foot } />
</table> `

File: packages/jsts/src/rules/helpers/table.ts (hunk 4 above)
Comment (vdiez) [reply packages/jsts/src/rules/helpers/table.ts]: got it, thx

File: packages/jsts/src/rules/helpers/table.ts [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+  if (rows.length === 0) {
+    return [];
+  }
+  const nbColumns = rows[0].length;
+  const columns: (TableCellInternal | undefined)[] = Array.from({ length: nbColumns });
+  let row = 0;
+  const result: TableCell[][] = [];
+  while (row < rows.length) {
+    const resultRow: TableCell[] = [];
+    let rowIndex = 0;
```
[the remark this answers is not in the document]
Comment (zglicz) [reply packages/jsts/src/rules/helpers/table.ts]: Renamed to indexInRow

File: packages/jsts/src/rules/S5260/rule.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        const elementType = getElementType(context)(tree.openingElement);
+        if (elementType === 'table') {
+          verifyHeaderReferences(tree);
+        }
+      },
+    };
+  },
+};
+
+function compileBlockInfo(grid: TableCell[][]) {
```
[the remark this answers is not in the document]
Comment (zglicz) [reply packages/jsts/src/rules/S5260/rule.ts:108]: Added a comment block

**Review by vdiez — APPROVED** [review]

**Pull request merged**